File: vscode-plugin/src/actions.ts
export interface Action {
	type: string;
}

export class ActionsInfo{
	constructor(public projectPath: string, public actions: Action[]) {}
}

export class MoveCaret implements Action {
	type: string = 'MOVE_CARET';
	offset!: number;
}

export class OpenFile implements Action {
	type: string = 'OPEN_FILE';
	path!: string;
	text!: string;
}

export class PrintText implements Action {
	type: string = 'PRINT_TEXT';
	text!: string;
	completable!: boolean;
}

export class DeleteRange implements Action {
	type: string = 'DELETE_RANGE';
	begin!: number;
	end!: number;
	completable!: boolean;
}

export class CallCompletion implements Action {
	type: string = 'CALL_COMPLETION';
	prefix!: string;
	expectedText!: string;
	tokenType!: string;
}

export class FinishSession implements Action {
	type: string = "FINISH_SESSION";
}

File: interpretator/src/org/jb/cce/actions/Actions.kt
package org.jb.cce.actions

import org.jb.cce.uast.NodeProperties

sealed class Action(val type: ActionType) {
    enum class ActionType {
        MOVE_CARET, CALL_COMPLETION, FINISH_SESSION, PRINT_TEXT, DELETE_RANGE
    }
}

data class FileActions(val path: String, val checksum: String, val sessionsCount: Int, val actions: List<Action>)

data class MoveCaret(val offset: Int) : Action(ActionType.MOVE_CARET)
data class CallCompletion(val prefix: String, val expectedText: String, val nodeProperties: NodeProperties) : Action(ActionType.CALL_COMPLETION)
class FinishSession : Action(ActionType.FINISH_SESSION)
data class PrintText(val text: String, val completable: Boolean = false) : Action(ActionType.PRINT_TEXT)
data class DeleteRange(val begin: Int, val end: Int, val completable: Boolean = false) : Action(ActionType.DELETE_RANGE)